SWIG 4.2

AMICI installation fails with swig4.2 from brew (https://github.com/AMICI-dev/AMICI/actions): 

```
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/build/temp.macosx-11.7-x86_64-cpython-39_amici/swig/CMakeFiles/_amici.dir/amiciPYTHON_wrap.cxx:53042:42: error: no matching constructor for initialization of 'amici::IDASolver'
          result = (amici::IDASolver *)new amici::IDASolver((amici::Solver const &)*arg1);
                                           ^                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/amici/swig/../include/amici/solver_idas.h:28:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const amici::Solver' to 'const amici::IDASolver' for 1st argument
  class IDASolver : public Solver {
        ^
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/amici/swig/../include/amici/solver_idas.h:28:7: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/build/temp.macosx-11.7-x86_64-cpython-39_amici/swig/CMakeFiles/_amici.dir/amiciPYTHON_wrap.cxx:54215:44: error: no matching constructor for initialization of 'amici::CVodeSolver'
          result = (amici::CVodeSolver *)new amici::CVodeSolver((amici::Solver const &)*arg1);
                                             ^                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/amici/swig/../include/amici/solver_cvodes.h:31:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const amici::Solver' to 'const amici::CVodeSolver' for 1st argument
  class CVodeSolver : public Solver {
        ^
  /amici_cdc3ecbbb3d74fdeb49899aa141c8808/amici/swig/../include/amici/solver_cvodes.h:31:7: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
```


swig4.2 *should* have been used already, e.g. by test_doc.yml since #2268. Double-check that we are specifying the swig executable correctly (`SWIG` vs `SWIG_EXECUTABLE` vs ...).

--- a/scripts/downloadAndBuildSwig.sh
+++ b/scripts/downloadAndBuildSwig.sh
@@ -18,13 +18,13 @@ cd "${AMICI_PATH}/ThirdParty/"
 
 if [[ ! -d ${SWIG_DIR} ]]; then
     if [[ ! -f ${SWIG_ARCHIVE} ]]
-        then wget ${SWIG_URL}
+        then wget "${SWIG_URL}"
     fi
 
-    tar -xzf ${SWIG_ARCHIVE}
+    tar -xzf "${SWIG_ARCHIVE}"
 fi
 
-cd ${SWIG_DIR}
+cd "${SWIG_DIR}"
 ./configure \
   --prefix="${PREFIX}" \
   --without-alllang \
@@ -38,6 +38,6 @@ echo "================"
 echo "SWIG installation successful"
 echo
 echo "To use this version of SWIG, add directory ${SWIG_BIN_DIR} to your PATH,"
-echo "e.g. adding the following line to your .bashrc:"
+echo "e.g., adding the following line to your ~/.bashrc:"
 echo "    export PATH=${SWIG_BIN_DIR}:\$PATH"
 echo "================"

--- a/swig/solver_cvodes.i
+++ b/swig/solver_cvodes.i
@@ -11,6 +11,8 @@ using namespace amici;
 
 // Required with SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
 %extend amici::CVodeSolver {
+    CVodeSolver() : Solver() {}
+
     CVodeSolver(Solver const& solver) {
         return new CVodeSolver(dynamic_cast<CVodeSolver const&>(solver));
     }

--- a/swig/solver_cvodes.i
+++ b/swig/solver_cvodes.i
@@ -11,7 +11,9 @@ using namespace amici;
 
 // Required with SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
 %extend amici::CVodeSolver {
-    CVodeSolver() : Solver() {}
+    CVodeSolver() {
+        return new CVodeSolver();
+    }
 
     CVodeSolver(Solver const& solver) {
         return new CVodeSolver(dynamic_cast<CVodeSolver const&>(solver));

--- a/swig/solver_cvodes.i
+++ b/swig/solver_cvodes.i
@@ -9,5 +9,12 @@ using namespace amici;
 %newobject amici::CVodeSolver::clone;
 %feature("notabstract") amici::CVodeSolver;
 
+// Required with SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
+%extend amici::CVodeSolver {
+    CVodeSolver(Solver const& solver) {
+        return new CVodeSolver(dynamic_cast<CVodeSolver const&>(solver));
+    }
+}
+
 // Process symbols in header
 %include "amici/solver_cvodes.h"

--- a/swig/solver_idas.i
+++ b/swig/solver_idas.i
@@ -11,6 +11,8 @@ using namespace amici;
 
 // Required for SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
 %extend amici::IDASolver {
+    IDASolver() : Solver() {}
+
     IDASolver(Solver const& solver) {
         return new IDASolver(dynamic_cast<IDASolver const&>(solver));
     }

--- a/swig/solver_idas.i
+++ b/swig/solver_idas.i
@@ -9,5 +9,12 @@ using namespace amici;
 %newobject amici::IDASolver::clone;
 %feature("notabstract") amici::IDASolver;
 
+// Required for SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
+%extend amici::IDASolver {
+    IDASolver(Solver const& solver) {
+        return new IDASolver(dynamic_cast<IDASolver const&>(solver));
+    }
+}
+
 // Process symbols in header
 %include "amici/solver_idas.h"

--- a/swig/solver_idas.i
+++ b/swig/solver_idas.i
@@ -11,7 +11,9 @@ using namespace amici;
 
 // Required for SWIG 4.2.0 https://github.com/AMICI-dev/AMICI/issues/2275
 %extend amici::IDASolver {
-    IDASolver() : Solver() {}
+    IDASolver() {
+        return new IDASolver();
+    }
 
     IDASolver(Solver const& solver) {
         return new IDASolver(dynamic_cast<IDASolver const&>(solver));

--- a/tests/cpp/unittests/testMisc.cpp
+++ b/tests/cpp/unittests/testMisc.cpp
@@ -261,6 +261,13 @@ TEST(SolverIdasTest, DefaultConstructableAndNotLeaky)
     IDASolver solver;
 }
 
+TEST(SolverIdasTest, CopyCtor)
+{
+    IDASolver solver1;
+    IDASolver solver2(solver1);
+}
+
+
 
 class SolverTest : public ::testing::Test {
   protected:

